### PR TITLE
Review request - Social media web app

### PR DESCRIPTION
**Pull request by ahmetbozaci**
**Milestone 2 - Friendship v1**

- Create model association for friendship
- Implement user profile to show
- Pending Friends
- Friends
- Show all user

### CODE REVIEW

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @ahmetbozaci & @chrystalme**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [x] please make sure to add unit and integration tests as you didn't add any.
- [x] users can only see posts that are created by them or users who are friends. right now a user can see a post created by a non-friend user.
- [x] when I click on invite to friendship I can't see the invite anywhere. not in the users#index page nor in the user#show page.
- [x] when I click on invite to friends, the friendship status should be converted to pending and the button should be removed. if the friendship is accepted you should display "friend" next to the user's name. if not, you should add the button again. 


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -15,4 +15,12 @@ def like_or_dislike_btn(post)
...
     end
   end
+
+  # def accept_or_deny_btn(friendship)
+  #   friend_request = Friendship.find_by( user: current_user , friendship: friendship)
+  #   if friend_request
+  #     link_to( 'Accept', user_friendship_path(@user, i.id))
+  #     link_to( 'Deny', user_friendships_path(@user))
+  #   end
+  # end
```
Comment (acushlakoncept) [inline app/helpers/application_helper.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

File: app/models/user.rb [hunk 2]
```diff
@@ -6,7 +6,41 @@ class User < ApplicationRecord
...
+    friends_array += inverse_friendships.map { |friendship| friendship.user if friendship.status }
+    friends_array.compact
+  end
+
+  # Users who have yet to confirmed friend requests
+  def pending_friends
+    friendships.map { |friendship| friendship.friend unless friendship.status }.compact
+  end
+
+  # Users who have requested to be friends
```
Comment (acushlakoncept) [inline app/models/user.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -1,41 +1,48 @@
...
-      <div class="notice">
-        <p> <%= notice %> </p>
+      <div class='menu-item'>
+        <%= menu_link_to 'All users', users_path %>
       </div>
+    </div>
+    <% if current_user %>
+    <%= link_to current_user.name, user_path(current_user) %> |
+    <%#= gravatar_image_tag(current_user.email, :gravatar => { :size => 20 }, alt: current_user.name) %>
+    <%#= link_to gravatar_image_tag(current_user.email, gravatar: { :size => 15 }, alt: current_user.name) %>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -1,41 +1,48 @@
...
+  </nav>
+  <% if notice.present? %>
+  <div class="notice">
+    <p> <%= notice %> </p>
+  </div>
+  <% end %>
+  <% if alert.present? %>
+  <div class="alert">
+    <p> <%= alert %> </p>
+  </div>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb]: There should be absolutely NO logic in views.

File: app/views/users/show.html.erb [hunk 5]
```diff
@@ -1,9 +1,41 @@
...
+    <% current_user.friends.each do |friend| %>
+    <li><%= friend.name %></li>
+    <% end %>
+    <% end %>
+  </ul>
+  <h3 class="text-danger">Pending Friends:</h3>
+  <% if current_user.pending_friendships.none? %>
+  <p>You have no pending friendships </p>
+  <% else %>
+  <% current_user.pending_friendships.each do |pf| %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb]: There should be absolutely NO logic in views.

File: app/views/users/profile.html.erb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+<%= link_to 'Deny', user_friendships_path(@user) , class:"btn btn-outline-secondary" %>
+<% else %>
+<p class="px-2">
+  <%= link_to 'Invite to Friend', user_friendships_path(@user, current_user), class:"btn btn-secondary", method: :post %>
+</p>
+<% end %>
+
+<% else %>
+<p class="btn btn-info disabled"> Friend </p>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/users/profile.html.erb]: There should be absolutely NO logic in views.

File: app/views/users/_user.html.erb [hunk 7]
```diff
@@ -1,6 +1,28 @@
...
+      <%= link_to 'Deny', user_friendships_path(user) , class:"btn btn-outline-secondary" %>
+    <% else %>
+      <p class="px-2">
+        <%= link_to 'Invite to Friend', user_friendships_path(user, current_user), class:"btn btn-secondary", method: :post %>
+      </p>
+    <% end %>
+  
+  <% else %>
+  <p class="btn btn-info disabled"> Friend </p>
+  <% end %>
```
Comment (acushlakoncept) [inline app/views/users/_user.html.erb]: There should be absolutely NO logic in views.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @ahmetbozaci and @chrystalme, Great job! 👏  Both of you have done well on this project.

> Great codebase 👍 

However, there are still some changes that are necessary for this project to be approved.
Let us quickly add some finishing touches and request another review.

- [x] Your posts timeline should include posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)), Currently I can see post from users that are not logged-in user friend.

- [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase see affected files below:

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

- [x] I see you have nicely implemented unit tests on all your model, kindly add integration tests as well.  [Capybara](https://github.com/teamcapybara/capybara) is an exceptionally great tool for performing an integration test with RSpec because it helps you perform end-to-end tests on your applications. It does this by simulating how a real user would interact with your application. Want to know more? See this [article](https://www.microverse.org/blog/understanding-rspec-controller-request-specs-integration-test-using-capybara). 
- [x] Don't forget to take care of all your pending tests.

- [x] Your ReadMe file still carries the app template description. "This repo includes initial code for social media app with basic styling. Its purpose is to be a starting point for Microverse students." I am sure you will agree with me, this repo no longer carries initial code, you have done a lot. Do well to update it to reflect what the app is all about. Possibly a screenshot too, to give it that professional look.


### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by BrianSammit — APPROVED** [review]
STATUS: APPROVED :green_circle: :1st_place_medal: 
-------------
Hello, @ahmetbozaci @chrystalme You did an excellent job on this milestone. Good job working on the required changes :muscle: I really like how the project looks. :1st_place_medal: :100: .
---------------
You can merge this milestone now :smile: 
Reach me on SLACK if you have any question 
@ Brian Sammit Cruz Rodriguez 
HAPPY CODING!! AND GOOD LUCK WITH THR NEXT MILESTONES. :crossed_fingers: :four_leaf_clover: 

Well-Done-Applause

**Pull request merged**